rules/prefer-key-over-key-code.js

```diff
@@ -0,0 +1,189 @@
+	}, null);
+};
+
+const isPropertyOf = (node, objectName) => {
+	return (
+		node &&
+		node.parent &&
+		node.parent.object &&
+		node.parent.object.name === objectName
+	);
+};
+
+const getEventNode = node => {
+	switch (node && node.type) {
+		case 'ArrowFunctionExpression':
+		case 'FunctionExpression': {
+			return node.params && node.params[0] && node.params[0].name;
+		}
+
+		default:
+			return null;
+	}
+};
+
+const extraCitation = 'See https://goo.gl/cRK532 for more info.';
+
+const util = context => {
+	const getNearestAncestorByType = nodes => type => {
+		for (let i = nodes.length - 1; i >= 0; i--) {
+			if (nodes[i].type === type) {
+				return nodes[i];
+			}
+		}
+	};
+
+	const report = reportError(context);
+	const ancestors = context.getAncestors();
+	const callExpressionNode = isInsideEventCallback(
+		ancestors.filter(n => n.type === 'CallExpression')
+	);
+	const event = getEventNode(
+		callExpressionNode && callExpressionNode.arguments[1]
+	);
+	return {
+		report,
+		event,
+		getNearestAncestorByType: getNearestAncestorByType(ancestors)
+	};
+};
+
+const directAccessRule = context => node => {
+	const {report, event} = util(context);
+	// Normal case when usage is direct -> event.keyCode
+
+	const isPropertyOfEvent = isPropertyOf(node, event);
+	if (isPropertyOfEvent) {
+		report(node);
+	}
+};
+
+const destructuredPropertyRule = context => node => {
+	const propertyName = node.value && node.value.name;
+	if (!keys.includes(propertyName)) {
+		return;
+	}
+
+	const {report, event, getNearestAncestorByType} = util(context);
+
+	// Destructured case
+	const nearestVariableDeclarator = getNearestAncestorByType(
+		'VariableDeclarator'
+	);
+	const isDestructuredFromEvent =
+		nearestVariableDeclarator &&
+		nearestVariableDeclarator.init &&
+		nearestVariableDeclarator.init.name === event;
+
+	if (isDestructuredFromEvent) {
+		report(node.value);
+	}
+};
+
+const fix = (context, node) => fixer => {
+	const {getNearestAncestorByType} = util(context);
+	const nearestIf = getNearestAncestorByType('IfStatement');
+	if (!nearestIf) {
+		return;
+	}
+
+	const {right, operator} = nearestIf.test;
+	const isTestingEquality = operator === '==' || operator === '===';
+	const isRightValid =
+		isTestingEquality && right.type === 'Literal' && typeof right.value === 'number';
+	// Either a meta key or a printable character
+	const keyCode = translateToKey[right.value] || String.fromCharCode(right.value);
+	// And if we recognize the keyCode
+	if (!isRightValid || !keyCode) {
+		return;
+	}
+
+	// Apply fixes
+	return [
+		fixer.replaceText(node, 'key'),
+		fixer.replaceText(right, `'${keyCode}'`)
+	];
+};
+
+const reportError = context => node => {
+	context.report({
+		message: `Use key instead of ${node.name}. ${extraCitation}`,
+		node,
+		fix: fix(context, node)
+	});
+};
+
+const create = context => {
+	return {
+		'Identifier:matches([name=keyCode], [name=charCode], [name=which])': directAccessRule(
```

Comment:
Why not inline the `directAccessRule` here, just like in all the other rules? This way readers won't have to jump back and forth to see which selector this visitor method is handling.

Reply:
Done!